osu.Game.Rulesets.Osu/Scoring/OsuHealthProcessor.cs

```diff
@@ -0,0 +1,229 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Linq;
+using osu.Game.Beatmaps;
+using osu.Game.Beatmaps.Timing;
+using osu.Game.Rulesets.Judgements;
+using osu.Game.Rulesets.Objects;
+using osu.Game.Rulesets.Osu.Objects;
+using osu.Game.Rulesets.Scoring;
+
+namespace osu.Game.Rulesets.Osu.Scoring
+{
+    public partial class OsuHealthProcessor : DrainingHealthProcessor
+    {
+        public Action<string>? OnIterationFail;
+        public Action<string>? OnIterationSuccess;
+
+        private double lowestHpEver;
+        private double lowestHpEnd;
+        private double hpRecoveryAvailable;
+        private double hpMultiplierNormal;
+
+        public OsuHealthProcessor(double drainStartTime)
+            : base(drainStartTime)
+        {
+        }
+
+        public override void ApplyBeatmap(IBeatmap beatmap)
+        {
+            lowestHpEver = IBeatmapDifficultyInfo.DifficultyRange(beatmap.Difficulty.DrainRate, 0.975, 0.8, 0.3);
+            lowestHpEnd = IBeatmapDifficultyInfo.DifficultyRange(beatmap.Difficulty.DrainRate, 0.99, 0.9, 0.4);
+            hpRecoveryAvailable = IBeatmapDifficultyInfo.DifficultyRange(beatmap.Difficulty.DrainRate, 0.04, 0.02, 0);
+
+            base.ApplyBeatmap(beatmap);
+        }
+
+        protected override void Reset(bool storeResults)
+        {
+            hpMultiplierNormal = 1;
+            base.Reset(storeResults);
+        }
+
+        protected override double ComputeDrainRate()
+        {
+            double testDrop = 0.00025;
+            double currentHp;
+            double currentHpUncapped;
+
+            while (true)
+            {
+                currentHp = 1;
+                currentHpUncapped = 1;
+
+                double lowestHp = currentHp;
+                double lastTime = DrainStartTime;
+                int currentBreak = 0;
+                bool fail = false;
+
+                for (int i = 0; i < Beatmap.HitObjects.Count; i++)
+                {
+                    HitObject h = Beatmap.HitObjects[i];
+
+                    // Find active break (between current and lastTime)
+                    double localLastTime = lastTime;
+                    double breakTime = 0;
+
+                    // TODO: This doesn't handle overlapping/sequential breaks correctly (/b/614).
+                    // Subtract any break time from the duration since the last object
+                    if (Beatmap.Breaks.Count > 0 && currentBreak < Beatmap.Breaks.Count)
```

Comment:
This logic is shockingly written, but I'm not going to touch it as it does match the stable code.

Reply:
See https://github.com/ppy/osu/pull/25420/files#diff-e868bcde9a3460149c8185e5bf30307c120a30c0f7db5354cf905e8c5f6c7b98R65-R72 for how this gets updated.